nested aux-build in tests/rustdoc/ tests

EtomicBomb opened the pull request:
* Fixes bug that prevented using nested aux-build in `tests/rustdoc/` tests. Before, `fn document` and the auxiliary builder disagreed about where to find the nested aux-build source file (`auxiliary/auxiliary/aux.rs` vs `auxiliary/aux.rs`), preventing them from building. Picked the latter in line with other builders in compiletest.
* Adds `//@ doc-flags` header, which forwards flags to rustdoc and not rustc. 
* Adds `//@ unique-doc-out-dir` header, which sets the --out-dir for the rustdoc invocation to a unique directory: `<root out dir>/docs/<test name>/doc`
* Changes working directory of the rustdoc invocation to the root out directory (common among all aux-builds). Prior art: exec_compiled_test in runtest.rs
* Adds tests that use nested aux builds and new headers

These changes provide useful capabilities for writing rustdoc tests on their own. They are also needed to test the implementation for the [mergable-rustdoc-cross-crate-info](https://github.com/rust-lang/rfcs/pull/3662) RFC.

try-job: x86_64-msvc

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r​? <reviewer name>
-->
<!-- homu-ignore:end -->

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10084340953/job/27882943961) [(plain)](https://github.com/rust-lang/rust/commit/44f18f6e125b5473a02322ebd5735e920cfb06fa/checks/27882943961/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.880 Building wheels for collected packages: reuse
#13 2.881   Building wheel for reuse (pyproject.toml): started
#13 3.127   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.128   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.129   Stored in directory: /tmp/pip-ephem-wheel-cache-uiqslxgq/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.131 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.528 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.529 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.058 Collecting virtualenv
#13 4.058 Collecting virtualenv
#13 4.109   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 4.300      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 30.2 MB/s eta 0:00:00
#13 4.358 Collecting platformdirs<5,>=3.9.1
#13 4.365   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.403 Collecting filelock<4,>=3.12.2
#13 4.431 Collecting distlib<1,>=0.3.7
#13 4.439   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.450      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 53.4 MB/s eta 0:00:00
#13 4.450      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 53.4 MB/s eta 0:00:00
#13 4.531 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.724 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      225216 kB
DirectMap2M:    10260480 kB
DirectMap1G:     8388608 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/6106b05b27988f4b946d7af219a6db95fb4477a1/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-6106b05b27988f4b946d7af219a6db95fb4477a1-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
fmt: checked 5376 files
tidy check
tidy: Skipping binary file check, read-only filesystem
tidy error: /checkout/tests/rustdoc/cross-crate-info/kitchen-sink/i.rs: too many trailing newlines (3)
##[error]tidy error: /checkout/tests/rustdoc/cross-crate-info/working-dir-examples/q.rs:8: line longer than 100 chars
##[error]tidy error: /checkout/tests/rustdoc/cross-crate-info/cargo-two/e.rs:17: line longer than 100 chars
##[error]tidy error: /checkout/tests/rustdoc/cross-crate-info/cargo-transitive-no-index/s.rs:13: line longer than 100 chars
##[error]tidy error: /checkout/tests/rustdoc/cross-crate-info/cargo-transitive/s.rs:20: line longer than 100 chars
##[error]tidy error: /checkout/src/tools/compiletest/src/command-list.rs:13: line not in alphabetical order
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.1)
  Downloading pip-24.1.2-py3-none-any.whl.metadata (3.6 kB)
Downloading pip-24.1.2-py3-none-any.whl (1.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 44.3 MB/s eta 0:00:00
Installing collected packages: pip
---
All checks passed!
checking C++ file formatting
some tidy checks failed

Command LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustc" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "4" "--extra-checks=py:lint,cpp:fmt" (failure_mode=DelayFail, stdout_mode=Print, stderr_mode=Print) did not execute successfully.
Created at: src/core/build_steps/tool.rs:1109:23
Executed at: src/core/build_steps/test.rs:1084:29

Build completed unsuccessfully in 0:01:18
```

</details>

EtomicBomb committed:
initial implementation of rustdoc nested aux-build

compiler-errors commented:
This PR's history should be cleaned quite a lot, at least rebasing out all of the merge commits but also probably the useless commits that represent intermediate work states like "code from this morning".

r? rustdoc

EtomicBomb committed:
ordering and wrapping cross-crate-info tests

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10084791648/job/27884344887) [(plain)](https://github.com/rust-lang/rust/commit/dc024fe987442f536b56773c38d15ca162240382/checks/27884344887/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.793 Building wheels for collected packages: reuse
#13 2.794   Building wheel for reuse (pyproject.toml): started
#13 3.040   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.041   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.042   Stored in directory: /tmp/pip-ephem-wheel-cache-ibq8_on_/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.044 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.432 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.433 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.950 Collecting virtualenv
#13 3.950 Collecting virtualenv
#13 3.996   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 4.206      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 27.4 MB/s eta 0:00:00
#13 4.266 Collecting filelock<4,>=3.12.2
#13 4.273   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.293 Collecting distlib<1,>=0.3.7
#13 4.299   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.310      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 51.2 MB/s eta 0:00:00
#13 4.343 Collecting platformdirs<5,>=3.9.1
#13 4.353   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.432 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.617 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.7s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      206784 kB
DirectMap2M:     8181760 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/6106b05b27988f4b946d7af219a6db95fb4477a1/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-6106b05b27988f4b946d7af219a6db95fb4477a1-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
fmt: checked 5376 files
tidy check
tidy: Skipping binary file check, read-only filesystem
tidy error: /checkout/tests/rustdoc/cross-crate-info/kitchen-sink/i.rs: too many trailing newlines (3)
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.1)
  Using cached pip-24.1.2-py3-none-any.whl.metadata (3.6 kB)
Using cached pip-24.1.2-py3-none-any.whl (1.8 MB)
Installing collected packages: pip
  Attempting uninstall: pip
---
All checks passed!
checking C++ file formatting
some tidy checks failed

Command LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustc" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "4" "--extra-checks=py:lint,cpp:fmt" (failure_mode=DelayFail, stdout_mode=Print, stderr_mode=Print) did not execute successfully.
Created at: src/core/build_steps/tool.rs:1109:23
Executed at: src/core/build_steps/test.rs:1084:29

Build completed unsuccessfully in 0:01:19
```

</details>

EtomicBomb committed:
reformatted rustdoc/cross-crate-info, fixing trailing newline issue

EtomicBomb commented:
Rebased the history! This is my first PR, so I'm just learning the ropes!

notriddle reviewed (commented):
AFAICT, this seems fine. In particular, I love these test cases!

I'm not as familiar with the deep guts of compiletest, though.

CC @jieyouxu

jieyouxu commented:
> I'm not as familiar with the deep guts of compiletest, though.
> 
> CC @jieyouxu

I'll try to take a look tomorrow for the compiletest changes.

EtomicBomb committed:
file_stem and comment per notriddle

bors commented:
:umbrella: The latest upstream changes (presumably #125443) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jieyouxu commented on src/tools/compiletest/src/runtest.rs:
```diff
@@ -1587,13 +1588,15 @@ impl<'test> TestCx<'test> {
     }
 
-    fn document(&self, out_dir: &Path) -> ProcRes {
+    /// `root_out_dir` and `root_testpaths` refer to the parameters of the actual test being run.
+    /// Auxiliaries, no matter how deep, have the same root_out_dir and root_testpaths.
+    fn document(&self, root_out_dir: &Path, root_testpaths: &TestPaths) -> ProcRes {
```
Question: do rustdoc tests support revisions like ui tests, if so, would this be a problem with revisioned rustdoc tests?

jieyouxu replied:
> I’ve never seen them used, and htmldocck itself doesn’t know about them, so I guess not.

Actually it's very easy:

https://github.com/rust-lang/rust/blob/1ddedbaa5919b7b3e70d984660e21e844c615c97/src/tools/compiletest/src/runtest.rs#L2661-L2662

no revisions are not supported for `tests/rustdoc` 😄

jieyouxu reviewed (approved):
Sorry for the wait, I just got to this PR today. The compiletest changes LGTM for the most part, just a question about potential test revision interactions.

EtomicBomb committed:
merge conflicts; fix rebase duplicating imports

EtomicBomb commented:
I just rebased to fix some merge conflicts. I think the PR still has an extraneous S-waiting-on-author label that rustbot assigned before I removed the merge commits last week.

jieyouxu commented:
@rustbot review

EtomicBomb commented:
I just rebased because of https://github.com/rust-lang/rust/pull/125443, which I think means it needs another review!

jieyouxu reviewed (approved):
Thanks, r=me for the compiletest changes.

bors commented:
:pushpin: Commit f91da72cb75f03739985926c5e5561611f783b80 has been approved by `notriddle`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle f91da72cb75f03739985926c5e5561611f783b80 -->
<!-- homu: {"type":"Approved","sha":"f91da72cb75f03739985926c5e5561611f783b80","approver":"notriddle","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

EtomicBomb committed:
canonicalize path in another place to fix #128411

tmandry commented:
@bors try

bors commented:
:hourglass: Trying commit bd23e0eb2694752f573eafd0d894a2b16d10fd7b with merge 96627dcf52a6ca490d422d857b196241a02342a2...
<!-- homu: {"type":"TryBuildStarted","head_sha":"bd23e0eb2694752f573eafd0d894a2b16d10fd7b","merge_sha":"96627dcf52a6ca490d422d857b196241a02342a2"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10171766934/job/28135814891)
Build commit: 96627dcf52a6ca490d422d857b196241a02342a2 (`96627dcf52a6ca490d422d857b196241a02342a2`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10171766934/job/28135814891"},"merge_sha":"96627dcf52a6ca490d422d857b196241a02342a2"} -->

tmandry commented:
@bors r=notriddle

bors commented:
:pushpin: Commit bd23e0eb2694752f573eafd0d894a2b16d10fd7b has been approved by `notriddle`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle bd23e0eb2694752f573eafd0d894a2b16d10fd7b -->
<!-- homu: {"type":"Approved","sha":"bd23e0eb2694752f573eafd0d894a2b16d10fd7b","approver":"notriddle","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
